src/data-model-providers/codegen/EmberAttributeStorageImpl.cpp

```diff
@@ -0,0 +1,164 @@
+    //  - raw values are just bytes
+    return provider->WriteValue(path, value.data());
+}
+
+CHIP_ERROR EmberAttributeStorageImpl::Read(const ConcreteAttributePath & path, Buffer buffer)
+{
+    VerifyOrReturnError(buffer.size() > 0, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    AttributePersistenceProvider * provider = GetAttributePersistenceProvider();
+    const EmberAfAttributeMetadata fakeMetadata{
+        .defaultValue  = EmberAfDefaultOrMinMaxAttributeValue(static_cast<uint32_t>(0)),
+        .attributeId   = path.mAttributeId,
+        .size          = static_cast<uint16_t>(buffer.size()),
+        .attributeType = attributeType(buffer),
+        .mask          = 0,
+    };
+
+    switch (buffer.type())
+    {
+    case AttributeStorage::Buffer::Type::kPrimitive: {
+        MutableByteSpan raw{ reinterpret_cast<uint8_t *>(buffer.data()), buffer.size() };
+        // numbers are just read as-is
+        ReturnErrorOnFailure(provider->ReadValue(path, &fakeMetadata, raw));
+
+        // Read size MUST match expected size.
+        VerifyOrReturnError(raw.size() == fakeMetadata.size, CHIP_ERROR_INVALID_INTEGER_VALUE);
+        return CHIP_NO_ERROR;
+    }
+    case AttributeStorage::Buffer::Type::kMutableByteSpan: {
+        // raw data is a mutable byte buffer that we modify
+        auto raw = reinterpret_cast<MutableByteSpan *>(buffer.data());
+        return provider->ReadValue(path, &fakeMetadata, *raw);
+    }
+    case AttributeStorage::Buffer::Type::kBytesOneByteLength:
+    case AttributeStorage::Buffer::Type::kStringOneByteLength: {
+        MutableByteSpan raw{ reinterpret_cast<uint8_t *>(buffer.data()), buffer.size() };
+        ReturnErrorOnFailure(provider->ReadValue(path, &fakeMetadata, raw));
+
+        // first byte should be a valid size, If we read an empty value, empty it out
+        if (raw.empty())
+        {
+            *reinterpret_cast<uint8_t *>(buffer.data()) = 0;
+        }
+        else
+        {
+            // we have valid data, validate it is not out of bounds
+            VerifyOrReturnError(raw[0] + 1 <= fakeMetadata.size, CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+        }
+
+        return CHIP_NO_ERROR;
+    }
+    case AttributeStorage::Buffer::Type::kBytesTwoByteLength:
+    case AttributeStorage::Buffer::Type::kStringTwoByteLength: {
+        VerifyOrReturnError(buffer.size() > 1, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        MutableByteSpan raw{ reinterpret_cast<uint8_t *>(buffer.data()), buffer.size() };
+        ReturnErrorOnFailure(provider->ReadValue(path, &fakeMetadata, raw));
+
+        // first byte should be a valid size, If we read an empty value, empty it out
+        if (raw.empty())
+        {
+            *reinterpret_cast<uint8_t *>(buffer.data())       = 0;
+            *(reinterpret_cast<uint8_t *>(buffer.data()) + 1) = 0;
+        }
+        else
+        {
+            // we have valid data, validate it is not out of bounds
+            uint16_t len;
+            memcpy(&len, raw.data(), 2);
```

Comment:
This is wrong. The length is little-endian.  Instead of hand-rolling all this stuff, if you don't want to use ember-strings.h, please define equivalent helpers (e.g. on PascalString?) and use them here for both reading and riting the lengths.  And perhaps change ember-strings.cpp to use those?